hibernate-core/src/main/java/org/hibernate/event/spi/EventType.java

```diff
@@ -76,6 +78,41 @@
 	public static final EventType<PostCollectionRemoveEventListener> POST_COLLECTION_REMOVE = create( "post-collection-remove", PostCollectionRemoveEventListener.class );
 	public static final EventType<PostCollectionUpdateEventListener> POST_COLLECTION_UPDATE = create( "post-collection-update", PostCollectionUpdateEventListener.class );
 
+	/**
+	 * Add a new event type.
+	 *
+	 * @param name - name of the custom event
+	 * @param listenerClass - the base listener class or interface associated with the entity type
+	 * @param <T> - listenerClass
+	 * @return the custom {@link EventType}
+	 */
+	public synchronized static <T> EventType<T> addCustomEventType(String name, Class<T> listenerClass) {
+		if ( name == null || listenerClass == null ) {
+			throw new HibernateException( "Custom EventType name and associated class must be non-null." );
+		}
+		final EventType existingEventType = EVENT_TYPE_BY_NAME_MAP.get( name );
+		if ( existingEventType != null ) {
+			if ( listenerClass.equals( existingEventType.baseListenerInterface ) ) {
+				LOG.warn( "EventType [" + name + "] with listener Class ["
+								  + listenerClass + "] was added more than once." );
+				return existingEventType;
+			}
+			else {
+				throw new HibernateException(
+						"Could not add EventType [" + name + "] with listener Class ["
+								+ "]. An EventType with that name already exists with listener ["
+								+ listenerClass.getName()
+								+ "]."
+				);
+			}
+		}
+		else {
+			final EventType<T> eventType = create( name, listenerClass );
+			EVENT_TYPE_BY_NAME_MAP.put( name, eventType );
+			LOG.info( "Added custom EventType:  [" + name + "], ordinal=[" +eventType.ordinal + "], listener=[" + listenerClass + "]." );
```

Comment:
Is "INFO" level not a bit too noisy? I'd use "DEBUG", as most users won't need to know how this works behind the scenes.

Reply:
I didn't think there would be all that many custom `EventType`s added, so I didn't think much about noisiness. I do see your point though about hibernate-rx users.

I'll make this change.

Reply:
It seems a good example to guard against unnecessary logging performance cost. The string concatenation cost is always incurred regardless of whether the logging statement is to be run or not. IMHO, we should either
```
if ( LOG.isInfoEnabled() ) {
    LOG.info( "Added custom EventType:  [" + name + "], ordinal=[" +eventType.ordinal + "], listener=[" + listenerClass + "]." );
}
```
or
```
LOG.info( "Added custom EventType:  [%s], ordinal=["%d"], listener=["%s"].",  name, eventType.ordinal, listenerClass );
```
I prefer the latter for its much better readability.

Reply:
> I didn't think there would be all that many custom EventTypes added, so I didn't think much about noisiness. 

I agree that I don't expect many custom types to be added.

But even assuming that RX would add a single type, if it's started twice you'd get this log message - unless the code in RX can somehow figure out that registering its own type is not necessary.

It's not unlikely to start a framework twice in the same JVM: one might have multiple applications in a server, or multiple persistence units in a single application, or running integration tests which start a new Hibernate instance for each test.  Since it's a static map, we need to take such possibilities into account.

A possible alternative would be not use a static map for this purpose.

Reply:
I missed this in my last commit. I'll get it in the next round.

FWIW, I don't think this is going to have much of an impact on performance. I don't imagine there would be a lot of custom `EventType`s that would be added.

Also, I believe adding custom `EventType`s and listener registration should only happen when the `SessionFactory` is being built. This should have no effect on an application's operation.

Reply:
@Sanne, I changed it to {{LOG.debug}} in my most recent commit. I'll change the message as @NathanQingyangXu suggests.